lib/src/main/scala/spinal/lib/Stream.scala

```diff
@@ -1859,3 +1860,69 @@ class StreamTransactionExtender[T <: Data, T2 <: Data](
     io.first := (counter.io.value === 0) && counter.io.working
     io.working := counter.io.working
 }
+/*
+  Unpacks a Stream into a layout by bit position.
+ */
+object StreamUnpacker {
+  def apply[T <: Data](input: Stream[T], layout: List[(Data, Int)]): StreamUnpacker[T] = {
+    require(layout.nonEmpty)
+    val map_layout = layout.toMapLinked()
+    val unpacker = new StreamUnpacker[T](input.payloadType, map_layout)
+    unpacker.io.input << input
+    map_layout.keys.zip(unpacker.io.outputs) foreach { case(data, comp_out) =>
+      data.assignFromBits(comp_out.asBits)
+    }
+    unpacker
+  }
+}
+
+class StreamUnpacker[T <: Data](
+  dataType: HardType[T],
+  layout: mutable.LinkedHashMap[Data, Int]
```

Comment:
Hi ^^

passing Data as arguements of a Component is problematic, as that data would be in the parent component.

The best here i think is to extends Area, that way it is all good, you can directly pass instances of Data and access them.

Reply:
Is there a way to pass in the HardTypes for the Data? I couldn't figure out how to get those from regular Data. It isn't as easy, unlike getting a Stream's `.payloadType`

Reply:
It complexify things i would say. you could have layout: mutable.LinkedHashMap[HardType[Data], Int]

But overall, i think the best is to use Area instead of Component, and directly work / assign the Data parameters.

Reply:
I gave up on the HardType map. I couldn't make a Vec of HardTypes but Data works fine.

I am going to move forward with Area instead of Component. Thanks for the suggestion!